src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -142,47 +142,111 @@ public string FullName
         {
             get
             {
-                // This is a recursive method over potentially hostile input. Protection against DoS is offered
-                // via the [Try]Parse method and TypeNameParserOptions.MaxNodes property at construction time.
-                // This FullName property getter and related methods assume that this TypeName instance has an
-                // acceptable node count.
-                //
-                // The node count controls the total amount of work performed by this method, including:
-                // - The max possible stack depth due to the recursive methods calls; and
-                // - The total number of bytes allocated by this function. For a deeply-nested TypeName
-                //   object, the total allocation across the full object graph will be
-                //   O(FullName.Length * GetNodeCount()).
-
-                if (_fullName is null)
+                if (_fullName is not null) // Fast path for already computed FullName and nested types
                 {
-                    if (IsConstructedGenericType)
+                    if (_nestedNameLength > 0 && _fullName.Length > _nestedNameLength) // Declaring types
                     {
-                        _fullName = TypeNameParserHelpers.GetGenericTypeFullName(GetGenericTypeDefinition().FullName.AsSpan(),
-#if SYSTEM_PRIVATE_CORELIB
-                            CollectionsMarshal.AsSpan(_genericArguments));
-#else
-                            _genericArguments.AsSpan());
-#endif
+                        // Stored fullName represents the full name of the nested type.
+                        // Example: Namespace.Declaring+Nested
+                        _fullName = _fullName.Substring(0, _nestedNameLength);
                     }
-                    else if (IsArray || IsPointer || IsByRef)
+
+                    return _fullName;
+                }
+
+                // We need a Stack<T> to avoid recursion, but use just List<T> as Stack<T> is not a part of CoreLib.
+                List<NameInfo> stack = [new(this, NameFlags.None)];
+                // Since we start from "this" which is the leaf node of the tree,
+                // we append all the information in a reverse order and then reverse the string at the end.
+                ValueStringBuilder builder = new(stackalloc char[128]);
+                while (stack.Count > 0)
+                {
+                    NameInfo currentInfo = stack[stack.Count - 1];
+                    stack.RemoveAt(stack.Count - 1);
```

Comment:
Decrease `stack.Count`? Otherwise it should be infinite loop.

Reply:
`RemoveAt` takes care of decreasing the `Count`